### PR TITLE
LSV: document hang reported in #37865

### PR DESCRIPTION
**Pull request by artagnon**
LoadStoreVectorizer hangs on certain examples, when its reorder function goes into a cycle. Detect this cycle and explicitly forbid it, using an assert, and document the resulting crash in a test-case under AArch64.

### CODE REVIEW

File: llvm/test/Transforms/LoadStoreVectorizer/AArch64/pr37865.ll [hunk 1]
```diff
@@ -0,0 +1,13 @@
+; REQUIRES: asserts
+; RUN: not --crash opt -mtriple=aarch64 -passes=load-store-vectorizer -disable-output %s
+
```
Comment (arsenm) [inline llvm/test/Transforms/LoadStoreVectorizer/AArch64/pr37865.ll:3]: should check the assert message

**Review by arsenm — COMMENTED** [review]
> It is unclear if this code-path will ever be hit when running the full opt pipeline.

Doesn't matter? Bugs are bugs? I think this pass is only added by the GPU backends in codegen

File: llvm/test/Transforms/LoadStoreVectorizer/AArch64/pr37865.ll (hunk 1 above)
Comment (arsenm) [reply llvm/test/Transforms/LoadStoreVectorizer/AArch64/pr37865.ll:3]: You added a dead check line and aren't running filecheck on stderr

File: llvm/test/Transforms/LoadStoreVectorizer/AArch64/pr37865.ll [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+define i32 @load_cycle(ptr %x) {
+; CHECK: Unexpected cycle while re-ordering instructions
+entry:
+  %gep.x.1 = getelementptr inbounds [2 x i32], ptr %x, i32 0, i32 1
+  %load.x.1 = load i32, ptr %gep.x.1
+  %rem = urem i32 %load.x.1, 1
+  %gep.x.2 = getelementptr inbounds [2 x i32], ptr %x, i32 %rem, i32 0
```
Comment (arsenm) [inline llvm/test/Transforms/LoadStoreVectorizer/AArch64/pr37865.ll:10]: I think this just failed to check that one of the vectorization pointer candidates actually depends on the other load, so these shouldn't be vectorization candidates

Comment (artagnon) [reply llvm/test/Transforms/LoadStoreVectorizer/AArch64/pr37865.ll:11]: Right, I figured that I have to fix the creation of the Chain.

**Review by arsenm — APPROVED** [review]

**Comment by bjope** [comment]
Hi @artagnon,

Downstream we have the fix from https://reviews.llvm.org/D54468, but it seems it never landed upstream.
The new test case you added here fails for us, since we do not trigger the new assert.

When looking at the review comments in https://reviews.llvm.org/D54468 it seems like there were some doubts about the validity of that patch. Nevertheless, we've used that fix downstream for nearly 6 years now.

**Comment by artagnon** [comment]
Hi @bjope,

Yes, I had a look at that patch, but I don't think it's the right fix unfortunately. Will attempt a proper fix in the coming days.

**Pull request merged**